File: src/examples/commits/parity.py
"""Cross-language parity twin of ``src/examples/commits/merkle-root.ts``.

The invariant this fixture proves: an EvidenceCommit's ``recordsRoot`` and
``hash`` are protocol values, not TypeScript implementation details. Given the
SAME pinned audit events, the SAME pinned envelope fields, and the SAME
manifest, the Python SDK must produce byte-identical digests to the TypeScript
SDK — because ``veritio-merkle-v1`` leaf/node hashing and ``veritio-commit-v1``
commit hashing both run over canonical JSON, and canonical JSON is a
cross-language byte contract.

Three things have to line up for the printed roots to agree, and each is a real
place the two SDKs could drift:

* ``hash_audit_record`` must canonicalize the record envelope exactly like
  ``hashAuditRecord``, or ``member.recordHash`` diverges before the tree is
  even built. The record hashes are printed for that reason: they localize a
  divergence to the record layer instead of the tree layer.
* ``create_evidence_commit`` must sort members by ``index`` rather than trusting
  the caller's array order. This fixture hands it the same two deliberately
  shuffled orderings the TypeScript fixture uses, so caller-order sensitivity
  would surface as ``orderIndependent: false`` here.
* ``_compute_records_root`` must duplicate the odd leaf at each level the same
  way. Three members is the smallest manifest that exercises that odd-leaf
  promotion, which is why the pinned manifest has three and not two or four.

Deliberate difference from the TypeScript expected output: there is NO
``recordChainVerification`` key. ``verifyAuditRecords`` is TypeScript-only —
the Python SDK exports ``verify_evidence_commits`` but no audit-record chain
verifier — so faking that field would assert a capability this SDK does not
have. Everything else is emitted under the same key names in the same order, so
the two expected files diff to exactly that one missing key.

Determinism: every id, timestamp, idempotency key and sequence is a literal.
``appendedAt`` in particular is pinned because a real conforming store stamps it
from wall-clock time, which would move every record hash — and therefore the
Merkle root — between runs.

Run with the sibling SDK on the path:

    PYTHONPATH=../veritio/sdks/python/src python3 src/examples/commits/parity.py
"""

import json
from typing import Any

from veritio import (
    create_audit_event,
    create_evidence_commit,
    hash_audit_record,
    hash_evidence_commit,
    hash_idempotency_key,
    verify_evidence_commits,
)

# Pinned verbatim from src/examples/commits/merkle-root.ts. These values exist
# only to be the same bytes the TypeScript fixture hashes; changing any of them
# invalidates the comparison even if both sides stay internally consistent.
COMMIT_SCOPE = {"tenantId": "org_acme", "environment": "production"}

MEMBERSHIP_EVENTS: list[dict[str, Any]] = [
    {
        "event": {
            "id": "evt_member_invited_01",
            "occurredAt": "2026-08-09T10:00:00.000Z",
            "actor": {"type": "user", "id": "usr_owner"},
            "action": "organization.member.invited",
            "target": {"type": "organization", "id": "org_acme"},
            "scope": COMMIT_SCOPE,
            "purpose": "access_management",
            "lawfulBasis": "contract",
            "retention": "security_1y",
            "metadata": {"role": "viewer"},
        },
        "idempotencyKey": "invitation:inv_123",
    },
    {
        "event": {
            "id": "evt_member_joined_01",
            "occurredAt": "2026-08-09T10:01:00.000Z",
            "actor": {"type": "user", "id": "usr_member"},
            "action": "organization.member.joined",
            "target": {"type": "organization", "id": "org_acme"},
            "scope": COMMIT_SCOPE,
            "purpose": "access_management",
            "lawfulBasis": "contract",
            "retention": "security_1y",
            "metadata": {"role": "viewer"},
        },
        "idempotencyKey": "membership:mem_123",
    },
    {
        "event": {
            "id": "evt_member_promoted_01",
            "occurredAt": "2026-08-09T10:02:00.000Z",
            "actor": {"type": "user", "id": "usr_owner"},
            "action": "organization.member.promoted",
            "target": {"type": "organization", "id": "org_acme"},
            "scope": COMMIT_SCOPE,
            "purpose": "access_management",
            "lawfulBasis": "contract",
            "retention": "security_1y",
            "metadata": {"role": "admin"},
        },
        "idempotencyKey": "promotion:mem_123",
    },
]

PINNED_APPENDED_AT = "2026-08-09T10:03:00.000Z"


def pinned_membership_records() -> list[dict[str, Any]]:
    """Build the three audit records the commit manifest binds.

    Mirrors ``pinnedMembershipRecords`` in the TypeScript fixture field for
    field. The chain (``sequence``, ``previousHash``) is built the way a
    conforming ``AuditStore`` builds it so ``member.recordHash`` references
    genuine envelope digests rather than placeholder bytes, but ``appendedAt``
    is pinned instead of clock-stamped: a moving ``appendedAt`` changes the
    record hash, which changes every Merkle leaf, which would make the parity
    claim unverifiable.
    """
    records: list[dict[str, Any]] = []
    previous_hash: str | None = None

    for position, entry in enumerate(MEMBERSHIP_EVENTS):
        record = {
            "event": create_audit_event(entry["event"]),
            "sequence": position + 1,
            "previousHash": previous_hash,
            "hashAlgorithm": "sha256",
            "canonicalization": "veritio-json-v1",
            "appendedAt": PINNED_APPENDED_AT,
            "idempotencyKeyHash": hash_idempotency_key(
                COMMIT_SCOPE["tenantId"], entry["idempotencyKey"]
            ),
        }
        record["hash"] = hash_audit_record(record)
        records.append(record)
        previous_hash = record["hash"]

    return records


def commit_members_for(records: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Project persisted records into commit manifest members.

    ``recordHash`` must be the algorithm-qualified ``sha256:<hex>`` form even
    though v1 audit record hashes are stored bare, and ``index`` — not list
    position — is the field ``create_evidence_commit`` canonicalizes and sorts
    on. Both rules are protocol, so both are reproduced here rather than left
    to the commit builder to guess.
    """
    return [
        {
            "index": position,
            "recordType": "audit.record",
            "recordId": record["event"]["id"],
            "recordHash": f"sha256:{record['hash']}",
        }
        for position, record in enumerate(records)
    ]


def commit_of(members: list[dict[str, Any]]) -> dict[str, Any]:
    """Commit a manifest in whatever list order the caller happens to hold it.

    ``commitId``, ``streamId``, ``sequence`` and ``committedAt`` are pinned to
    the TypeScript fixture's values because all four are canonicalized into the
    commit hash; leaving ``committedAt`` to default would stamp the current
    clock and break both determinism and the cross-language comparison.
    """
    return create_evidence_commit(
        {
            "commitId": "cmt_membership_01",
            "streamId": "org_acme:production",
            "sequence": 1,
            "previousCommitHash": None,
            "members": list(members),
            "committedAt": "2026-08-09T10:05:00.000Z",
        }
    )


def main() -> None:
    """Print the commit-parity report for the pinned membership manifest.

    Emits the same key names and nesting as ``merkle-root.ts`` (minus the
    TypeScript-only record-chain verification) so the two expected outputs can
    be diffed directly. Free of clock, randomness, environment and filesystem
    access, so stdout is byte-identical across runs and machines.
    """
    records = pinned_membership_records()
    members = commit_members_for(records)

    # Two deliberately different caller orderings of the same three members,
    # copied from the TypeScript fixture so both languages test the same
    # permutations rather than each picking a convenient one.
    shuffled = [members[2], members[0], members[1]]
    reversed_members = list(reversed(members))

    commit = commit_of(shuffled)
    rebuilt = commit_of(reversed_members)

    print(
        json.dumps(
            {
                "suppliedOrder": [
                    {"index": member["index"], "recordId": member["recordId"]}
                    for member in shuffled
                ],
                "canonicalMemberOrder": [
                    {
                        "index": member["index"],
                        "recordType": member["recordType"],
                        "recordId": member["recordId"],
                        "recordHash": member["recordHash"],
                    }
                    for member in commit["members"]
                ],
                "recordCount": commit["recordCount"],
                "treeAlgorithm": commit["treeAlgorithm"],
                "recordsRoot": commit["recordsRoot"],
                "commitHash": commit["hash"],
                "hashSelfConsistent": hash_evidence_commit(commit) == commit["hash"],
                "orderIndependent": rebuilt["hash"] == commit["hash"]
                and rebuilt["recordsRoot"] == commit["recordsRoot"],
                "commitVerification": verify_evidence_commits([commit]),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: src/examples/commits/parity.go
/*
Package main proves cross-language parity of the veritio-merkle-v1 evidence
commit tree.

This fixture rebuilds, with the Go SDK, the exact pinned commit that
src/examples/commits/merkle-root.ts builds with the TypeScript SDK: the same
three membership events, the same pinned appendedAt, the same tenant-scoped
idempotency keys, the same commit id / stream / sequence / committedAt. If the
canonical JSON, the audit-record envelope hash, the Merkle leaf/parent domain
separation, or the commit hash preimage drifted between languages, the printed
recordHash values, recordsRoot, or commitHash would differ from the TypeScript
expected output. Byte-identical values are the proof that they have not.

It also re-proves in Go the ordering invariant the TypeScript fixture proves:
EvidenceCommit member order is protocol-canonical, not caller-dependent, so two
callers handing the same manifest in different slice orders get the same root
and the same commit hash.

Scope note: chain verification over audit records (verifyAuditRecords) exists
only in the TypeScript SDK. Go implements VerifyEvidenceCommits, which proves
the commit ledger's internal consistency, so that is what this fixture asserts.
Every value below is pinned; the program is deterministic across runs.
*/
package main

import (
	"encoding/json"
	"fmt"

	veritio "github.com/getveritio/veritio/sdks/go"
)

// commitTenantID is the tenant scope every pinned record and idempotency key is
// bound to. It matches commitScope.tenantId in merkle-root.ts.
const commitTenantID = "org_acme"

// pinnedAppendedAt stands in for the wall-clock stamp a conforming AuditStore
// would write. It is pinned because it is hash-affecting: a live value would
// move every record hash between runs and defeat the parity comparison.
const pinnedAppendedAt = "2026-08-09T10:03:00.000Z"

// membershipEntry pairs one pinned audit event input with the host idempotency
// key the store would bind to it.
type membershipEntry struct {
	input          veritio.AuditEventInput
	idempotencyKey string
}

// commitScope is the shared evidence scope of the pinned membership stream.
func commitScope() *veritio.EvidenceScope {
	return &veritio.EvidenceScope{TenantID: commitTenantID, Environment: "production"}
}

// membershipEntries returns the same three membership events, in the same
// order and with the same field values, that the TypeScript fixture pins. Any
// divergence here invalidates the parity claim, so these literals must stay in
// lockstep with merkle-root.ts.
func membershipEntries() []membershipEntry {
	return []membershipEntry{
		{
			input: veritio.AuditEventInput{
				ID:          "evt_member_invited_01",
				OccurredAt:  "2026-08-09T10:00:00.000Z",
				Actor:       veritio.Principal{Type: "user", ID: "usr_owner"},
				Action:      "organization.member.invited",
				Target:      veritio.Resource{Type: "organization", ID: "org_acme"},
				Scope:       commitScope(),
				Purpose:     "access_management",
				LawfulBasis: "contract",
				Retention:   "security_1y",
				Metadata:    map[string]any{"role": "viewer"},
			},
			idempotencyKey: "invitation:inv_123",
		},
		{
			input: veritio.AuditEventInput{
				ID:          "evt_member_joined_01",
				OccurredAt:  "2026-08-09T10:01:00.000Z",
				Actor:       veritio.Principal{Type: "user", ID: "usr_member"},
				Action:      "organization.member.joined",
				Target:      veritio.Resource{Type: "organization", ID: "org_acme"},
				Scope:       commitScope(),
				Purpose:     "access_management",
				LawfulBasis: "contract",
				Retention:   "security_1y",
				Metadata:    map[string]any{"role": "viewer"},
			},
			idempotencyKey: "membership:mem_123",
		},
		{
			input: veritio.AuditEventInput{
				ID:          "evt_member_promoted_01",
				OccurredAt:  "2026-08-09T10:02:00.000Z",
				Actor:       veritio.Principal{Type: "user", ID: "usr_owner"},
				Action:      "organization.member.promoted",
				Target:      veritio.Resource{Type: "organization", ID: "org_acme"},
				Scope:       commitScope(),
				Purpose:     "access_management",
				LawfulBasis: "contract",
				Retention:   "security_1y",
				Metadata:    map[string]any{"role": "admin"},
			},
			idempotencyKey: "promotion:mem_123",
		},
	}
}

// pinnedMembershipRecords builds the three audit records the commit manifest
// binds, the way a conforming store builds a chain: sequence starts at 1,
// previousHash links to the prior record hash, and the envelope hash comes from
// the protocol helper rather than a placeholder. Only appendedAt is pinned.
func pinnedMembershipRecords() ([]veritio.AuditRecord, error) {
	entries := membershipEntries()
	records := make([]veritio.AuditRecord, 0, len(entries))
	var previousHash *string

	for position, entry := range entries {
		event, err := veritio.CreateAuditEvent(entry.input)
		if err != nil {
			return nil, err
		}
		idempotencyKeyHash, err := veritio.HashIdempotencyKey(commitTenantID, entry.idempotencyKey)
		if err != nil {
			return nil, err
		}

		record := veritio.AuditRecord{
			Event:              event,
			Sequence:           position + 1,
			PreviousHash:       previousHash,
			HashAlgorithm:      veritio.HashAlgorithm,
			Canonicalization:   veritio.Canonicalization,
			AppendedAt:         pinnedAppendedAt,
			IdempotencyKeyHash: idempotencyKeyHash,
		}
		hash, err := veritio.HashAuditRecord(record)
		if err != nil {
			return nil, err
		}
		record.Hash = hash
		records = append(records, record)

		linked := hash
		previousHash = &linked
	}

	return records, nil
}

// commitMembersFor projects persisted records into a commit manifest.
// recordHash must be the algorithm-qualified sha256:<hex> form even though v1
// audit record hashes are stored bare, and index — not slice position — is the
// field the protocol canonicalizes and sorts on.
func commitMembersFor(records []veritio.AuditRecord) []veritio.EvidenceCommitMember {
	members := make([]veritio.EvidenceCommitMember, 0, len(records))
	for position, record := range records {
		members = append(members, veritio.EvidenceCommitMember{
			Index:      position,
			RecordType: "audit.record",
			RecordID:   record.Event.ID,
			RecordHash: "sha256:" + record.Hash,
		})
	}
	return members
}

// commitOf commits a manifest in whatever slice order the caller happens to
// hold it. Commit id, stream, sequence, previousCommitHash, and committedAt are
// pinned to the TypeScript fixture's values so the Merkle root and commit hash
// are reproducible across runs and across languages.
func commitOf(members []veritio.EvidenceCommitMember) (veritio.EvidenceCommit, error) {
	ordered := make([]veritio.EvidenceCommitMember, len(members))
	copy(ordered, members)
	return veritio.CreateEvidenceCommit(veritio.EvidenceCommitInput{
		CommitID:           "cmt_membership_01",
		StreamID:           "org_acme:production",
		Sequence:           1,
		PreviousCommitHash: nil,
		Members:            ordered,
		CommittedAt:        "2026-08-09T10:05:00.000Z",
	})
}

// suppliedMember is the compact projection of a caller-supplied member used to
// show the deliberately non-canonical input ordering.
type suppliedMember struct {
	Index    int    `json:"index"`
	RecordID string `json:"recordId"`
}

// canonicalMember mirrors the member fields the protocol canonicalizes, so the
// printed manifest can be diffed field-for-field against the TypeScript output.
type canonicalMember struct {
	Index      int    `json:"index"`
	RecordType string `json:"recordType"`
	RecordID   string `json:"recordId"`
	RecordHash string `json:"recordHash"`
}

// parityReport fixes the output key order to match the TypeScript fixture's
// object literal, so the two expected files can be diffed directly. Go maps
// would serialize in alphabetical order and break that comparison.
type parityReport struct {
	SuppliedOrder        []suppliedMember                         `json:"suppliedOrder"`
	CanonicalMemberOrder []canonicalMember                        `json:"canonicalMemberOrder"`
	RecordCount          int                                      `json:"recordCount"`
	TreeAlgorithm        string                                   `json:"treeAlgorithm"`
	RecordsRoot          string                                   `json:"recordsRoot"`
	CommitHash           string                                   `json:"commitHash"`
	HashSelfConsistent   bool                                     `json:"hashSelfConsistent"`
	OrderIndependent     bool                                     `json:"orderIndependent"`
	CommitVerification   veritio.EvidenceCommitVerificationResult `json:"commitVerification"`
}

// main builds the pinned commit twice from two different caller orderings and
// prints the canonical manifest, the veritio-merkle-v1 recordsRoot, and the
// commit hash. The printed digests must equal the ones in
// src/examples/commits/expected/merkle-root.txt.
func main() {
	records, err := pinnedMembershipRecords()
	if err != nil {
		panic(err)
	}
	members := commitMembersFor(records)

	// Two deliberately different caller orderings of the same three members.
	shuffled := []veritio.EvidenceCommitMember{members[2], members[0], members[1]}
	reversed := []veritio.EvidenceCommitMember{members[2], members[1], members[0]}

	commit, err := commitOf(shuffled)
	if err != nil {
		panic(err)
	}
	rebuilt, err := commitOf(reversed)
	if err != nil {
		panic(err)
	}

	recomputed, err := veritio.HashEvidenceCommit(commit)
	if err != nil {
		panic(err)
	}

	supplied := make([]suppliedMember, 0, len(shuffled))
	for _, member := range shuffled {
		supplied = append(supplied, suppliedMember{Index: member.Index, RecordID: member.RecordID})
	}
	canonical := make([]canonicalMember, 0, len(commit.Members))
	for _, member := range commit.Members {
		canonical = append(canonical, canonicalMember{
			Index:      member.Index,
			RecordType: member.RecordType,
			RecordID:   member.RecordID,
			RecordHash: member.RecordHash,
		})
	}

	report := parityReport{
		SuppliedOrder:        supplied,
		CanonicalMemberOrder: canonical,
		RecordCount:          commit.RecordCount,
		TreeAlgorithm:        commit.TreeAlgorithm,
		RecordsRoot:          commit.RecordsRoot,
		CommitHash:           commit.Hash,
		HashSelfConsistent:   recomputed == commit.Hash,
		OrderIndependent:     rebuilt.Hash == commit.Hash && rebuilt.RecordsRoot == commit.RecordsRoot,
		CommitVerification:   veritio.VerifyEvidenceCommits([]veritio.EvidenceCommit{commit}),
	}

	encoded, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(encoded))
}
